SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -0,0 +1,560 @@
+  | 3 => BitVec.sle x y
+  | 4 => x.toInt > y.toInt
+  | 5 => x.toInt ≥  y.toInt
+  | 6 => BitVec.ult x y
+  | 7 => BitVec.ule x y
+  | 8 => x.toNat > y.toNat
+  | 9 => x.toNat > y.toNat
+  /- ceq/cne = SV case in/equality ("===")
+    tests 4-state logical in/equality (0, 1, x, z)
+    TODO: model the concept.
+  -/
+  | 10 => x == y
+  | 11 => x == y
+  /- weq/wne = SV wild card inequality ("==?")
+    also based on 4-state logical in/equality,
+    with x and z values acting as wild cards.
+    TODO: model the concept.
+  -/
+  | 12 => x == y
+  | 13 => x == y
+  | _ => false
+
+def mods (x y : BitVec w) : BitVec w :=
+  BitVec.smod x y
+
+def modu (x y : BitVec w) : BitVec w :=
+  BitVec.umod x y
+
+def mul (x y : BitVec w) : BitVec w :=
+  x * y
+
+/- according to the docs the input of mux
+  can be any type. I'll stick to bitvecs for now. -/
+def mux (x y : BitVec w) (cond : Bool) : BitVec w :=
+  match cond with
+  | true => x
+  | false => y
+
+def or (x y : BitVec w) : BitVec w :=
+  x ||| y
+
+def parity (x : BitVec w) : Bool :=
+  (BitVec.umod x 2#w) == 1
+
+def replicate (x : BitVec w) (n : Nat) : BitVec (w * n) :=
+  BitVec.replicate n x
+
+def shl (x y : BitVec w) : BitVec w :=
+  x <<< y
+
+def shrs (x y : BitVec w) : BitVec w :=
+  BitVec.sshiftRight' x y
+
+def shru (x y : BitVec w) : BitVec w :=
+  BitVec.ushiftRight x y.toNat
+
+def sub (x y : BitVec w) : BitVec w :=
+  x - y
+
+/-
+  Return true/false based on a lookup table.
+  This operation assumes a fully elaborated table – 2^n entries.
+  Inputs are sorted MSB -> LSB from left to right and the offset into lookupTable is computed from them.
+  The table is sorted from 0 -> (2^n - 1) from left to right.
+  No difference from array_get into an array of constants except for xprop behavior.
+  If one of the inputs is unknown, but said input doesn’t make a difference in the output (based on the lookup table)
+  the result should not be ‘x’ – it should be the well-known result.
+  Input size should also be variadic, but I am considering two inputs only for the time being.
+-/
+-- def truthTable (lookupTable : List Bool) (x y : BitVec 1) : BitVec 1 :=
+/-
+  I need to meditate a bit more on how to define this.
+-/
+
+def xor (x y : BitVec w) : BitVec w :=
+  x ^^^ y
+
+end Comb
+
+namespace MLIR2Comb
+
+section Dialect
+
+/-
+  One thing i've tried in these defs
+  is to give w (and potentially additional params, such as `n` in replicate and extract)
+  as additional type-like arguments. I believe this is not correct, however, I could not think of
+  any alternative solutions.
+-/
+
+inductive Op
+| add (w : Nat)
+| and (w : Nat)
+| concat (w₁ : Nat) (w₂ : Nat)
+| divs (w : Nat)
+| divu (w : Nat)
+| extract (w : Nat) (n : Nat) -- I tried to avoid the two args but could not think of a better solution
+| icmp (w : Nat) (n : Nat)
+| mods (w : Nat)
+| modu (w : Nat)
+| mul (w : Nat)
+| mux (w : Nat)
+| or (w : Nat)
+| parity (w : Nat)
+| replicate (w : Nat) (n : Nat)
```

Comment:
I think having the second argument here is fine; I presume it's needed because the output type of `replicate w n` is a bitvector of length `w * n`, and we don't allow dependent types in our model.

That does mean that we can only replicate by a static/constant amount. The [Circt docs](https://circt.llvm.org/docs/Dialects/Comb/#combreplicate-circtcombreplicateop) indeed describe `comb.replicate` as "Concatenate the operand a constant number of times", and specify only a single operand.